scout/commands/update/genes.py

```diff
@@ -17,94 +17,170 @@
+            resource_exists = os.path.isfile(resource_path)
+            if resource_exists:
+                resources[resname] = get_file_handle(resource_path).readlines()
+
+        # If the resource is manadatory make sure it exists and contains data (OMIM data is NOT mandatory)
+        if resname in ["hpo_genes", "hgnc_lines", "exac_lines"] and not resources.get(resname):
+            LOG.error(f"Missing resource {resname} in downloads path.")
+            raise click.Abort()
+
+        # Check that the available genes and transcripts file correspond to the required genome build
+        for build in builds:
+            if resname.endswith(build) and resources.get(resname) is False:
+                LOG.error(
+                    f"Updating genes for genome build '{build}' requires a resource '{resname}' that is currenly missing in provided path."
+                )
+                raise click.Abort()
+
+        # Check that resource lines contain actual data
+        if resname not in resources:
+            continue
+        if "<!DOCTYPE html>" in resources[resname][0] or "<!DOCTYPE html>" in resources[resname][1]:
+            LOG.error(f"Resource file '{resname}' doesn't contain valid data.")
+            raise click.Abort()
+
+
 @click.command("genes", short_help="Update all genes")
 @click.option(
     "--build",
     type=click.Choice(["37", "38"]),
     help="What genome build should be used. If no choice update 37 and 38.",
 )
-@click.option("--api-key", help="Specify the api key")
+@click.option(
+    "-f",
+    "--downloads_folder",
+    type=click.Path(exists=True, dir_okay=True, readable=True),
+    help="specify path to folder where files necessary to update genes are pre-downloaded",
+)
+@click.option(
+    "--api-key", help="Specify the OMIM downloads api key. Only if downloads_folder is not provided"
+)
 @with_appcontext
-def genes(build, api_key):
+def genes(build, downloads_folder, api_key):
     """
     Load the hgnc aliases to the mongo database.
     """
     LOG.info("Running scout update genes")
     adapter = store
-
-    # Fetch the omim information
+    builds = [build] if build else ["37", "38"]
     api_key = api_key or current_app.config.get("OMIM_API_KEY")
-    mim_files = {}
-    if not api_key:
-        LOG.warning("No omim api key provided, Please not that some information will be missing")
-
-    else:
-        try:
-            mim_files = fetch_mim_files(api_key, mim2genes=True, morbidmap=True, genemap2=True)
-        except Exception as err:
-            LOG.warning(err)
-            raise click.Abort()
+    resources = {}
+
+    # If required resources are missing, download them to a temporary file
+    if downloads_folder is None:
+        with tempfile.TemporaryDirectory() as tempdir:
+            try:
+                download_resources(tempdir, api_key, builds)
+            except Exception as ex:
+                LOG.error(ex)
+        fetch_downloaded_resources(resources, tempdir, builds)
```

Comment:
This is elegant and seems to work, but I don't understand why tempdir still lives outside the with: context, without being mentioned before. Something with the way the scope is dynamically computed there I guess, but it would be fun to know!

Reply:
I guess the temp dir is created in the file system outside the app, so it's not under the app context. The app context is instead just containing the stuff inside the scout/ folder.

To check you could print the path to the temp dir and the path to the app context

Reply:
Hm, yes, perhaps the thing is simply that with doesn't really set scope; the local scope here is the if clause and while inside it, the dir still exists, outside it is destroyed.